### ISSUE TITLE
Qwen_PI model question

### ISSUE THREAD

**Issue by fankeyu123** [starVLA/starVLA#217]
Thanks for your wonderful work!
I want to know when use Qwen_PI model, how to choose num_layers in diffusion_model_cfg(must it be the same as VLM hidden layers?).

**Comment by JinhuiYE** [comment]
yes

**Comment by fankeyu123** [comment]
how can you guarantee the efficiency(this may cost too many gpu memory)?



---Original---
From: "Jinhui ***@***.***&gt;
Date: Thu, Mar 26, 2026 18:12 PM
To: ***@***.***&gt;;
Cc: "Keyu ***@***.******@***.***&gt;;
Subject: Re: [starVLA/starVLA] Qwen_PI model question (Issue #217)


JinhuiYE left a comment (starVLA/starVLA#217)
 
yes

**Comment by JinhuiYE** [comment]
This is because when going from PI, the book to be changed can be set using
self.config.framework.qwenvl.num_vl_layers = num_vl_layers.

**Comment by fankeyu123** [comment]
then how about the training loss? how much is it normal to converge?

**Comment by JinhuiYE** [comment]
You can refer to the W&B loss curves shared in #68: https://wandb.ai/jinhuiye/StarVLA_Libero and the RoboTwin run at https://api.wandb.ai/links/axi-the-cat/7fus3mhz. The loss starts relatively low (not random-init scale) because of normalization layers on the action output and at the top of the DiT — see #68 for details. For Qwen-PI specifically, if the trajectories still look unstable, also check that you're using the latest code (with action output layer normalization) and increase denoising steps from the default 4 to 10+ as discussed in #212.

**Comment by fankeyu123** [comment]
thx a lot！



---Original---
From: "Jinhui ***@***.***&gt;
Date: Fri, Mar 27, 2026 12:22 PM
To: ***@***.***&gt;;
Cc: "Keyu ***@***.******@***.***&gt;;
Subject: Re: [starVLA/starVLA] Qwen_PI model question (Issue #217)


JinhuiYE left a comment (starVLA/starVLA#217)
 
You can refer to the W&amp;B loss curves shared in #68: https://wandb.ai/jinhuiye/StarVLA_Libero and the RoboTwin run at https://api.wandb.ai/links/axi-the-cat/7fus3mhz. The loss starts relatively low (not random-init scale) because of normalization layers on the action output and at the top of the DiT — see #68 for details. For Qwen-PI specifically, if the trajectories still look unstable, also check that you're using the latest code (with action output layer normalization) and increase denoising steps from the default 4 to 10+ as discussed in #212.

**Comment by wowfisch** [comment]
> You can refer to the W&B loss curves shared in [#68](https://github.com/starVLA/starVLA/issues/68): https://wandb.ai/jinhuiye/StarVLA_Libero and the RoboTwin run at https://api.wandb.ai/links/axi-the-cat/7fus3mhz. The loss starts relatively low (not random-init scale) because of normalization layers on the action output and at the top of the DiT — see [#68](https://github.com/starVLA/starVLA/issues/68) for details. For Qwen-PI specifically, if the trajectories still look unstable, also check that you're using the latest code (with action output layer normalization) and increase denoising steps from the default 4 to 10+ as discussed in [#212](https://github.com/starVLA/starVLA/issues/212).

@JinhuiYE "increase denoising steps from the default 4 to 10+" stands for num_inference_timesteps rather than repeated_diffusion_steps? thx

**Comment by fankeyu123** [comment]
> > You can refer to the W&B loss curves shared in [#68](https://github.com/starVLA/starVLA/issues/68): https://wandb.ai/jinhuiye/StarVLA_Libero and the RoboTwin run at https://api.wandb.ai/links/axi-the-cat/7fus3mhz. The loss starts relatively low (not random-init scale) because of normalization layers on the action output and at the top of the DiT — see [#68](https://github.com/starVLA/starVLA/issues/68) for details. For Qwen-PI specifically, if the trajectories still look unstable, also check that you're using the latest code (with action output layer normalization) and increase denoising steps from the default 4 to 10+ as discussed in [#212](https://github.com/starVLA/starVLA/issues/212).
> 
> [@JinhuiYE](https://github.com/JinhuiYE) "increase denoising steps from the default 4 to 10+" stands for num_inference_timesteps rather than repeated_diffusion_steps? thx

I think yes. Repeated_diffusion_steps is used to sample more for diffusion steps.

**Comment by wowfisch** [comment]
> > > You can refer to the W&B loss curves shared in [#68](https://github.com/starVLA/starVLA/issues/68): https://wandb.ai/jinhuiye/StarVLA_Libero and the RoboTwin run at https://api.wandb.ai/links/axi-the-cat/7fus3mhz. The loss starts relatively low (not random-init scale) because of normalization layers on the action output and at the top of the DiT — see [#68](https://github.com/starVLA/starVLA/issues/68) for details. For Qwen-PI specifically, if the trajectories still look unstable, also check that you're using the latest code (with action output layer normalization) and increase denoising steps from the default 4 to 10+ as discussed in [#212](https://github.com/starVLA/starVLA/issues/212).
> > 
> > 
> > [@JinhuiYE](https://github.com/JinhuiYE) "increase denoising steps from the default 4 to 10+" stands for num_inference_timesteps rather than repeated_diffusion_steps? thx
> 
> I think yes. Repeated_diffusion_steps is used to sample more for diffusion steps.

btw, have you found where is the implementation of action output layer normalization？i only find usage of layernorm in the code "hidden_states = self.norm_out(hidden_states) * (1 + scale[:, None]) + shift[:, None]" in dit but last update is 6 months ago, which is far away from "the latest code"

**Comment by fankeyu123** [comment]
there is no layernorm，and it affact training convergence, so I add it by myself



---Original---
From: ***@***.***&gt;
Date: Thu, Apr 9, 2026 11:25 AM
To: ***@***.***&gt;;
Cc: "Keyu ***@***.******@***.***&gt;;
Subject: Re: [starVLA/starVLA] Qwen_PI model question (Issue #217)


wowfisch left a comment (starVLA/starVLA#217)
    
You can refer to the W&amp;B loss curves shared in #68: https://wandb.ai/jinhuiye/StarVLA_Libero and the RoboTwin run at https://api.wandb.ai/links/axi-the-cat/7fus3mhz. The loss starts relatively low (not random-init scale) because of normalization layers on the action output and at the top of the DiT — see #68 for details. For Qwen-PI specifically, if the trajectories still look unstable, also check that you're using the latest code (with action output layer normalization) and increase denoising steps from the default 4 to 10+ as discussed in #212.
  
@JinhuiYE "increase denoising steps from the default 4 to 10+" stands for num_inference_timesteps rather than repeated_diffusion_steps? thx
  
I think yes. Repeated_diffusion_steps is used to sample more for diffusion steps.
  
btw, have you found where is the implementation of action output layer normalization？i only find usage of layernorm in the code "hidden_states = self.norm_out(hidden_states) * (1 + scale[:, None]) + shift[:, None]" in dit but last update is 6 months ago, which is far away from "the latest code"

**Comment by wowfisch** [comment]
@fankeyu123 you mean here `pred = self.action_decoder(model_output)`？

**Comment by JinhuiYE** [comment]
> there is no layernorm，and it affact training convergence, so I add it by myself
> […](#)

yep, because we have already released ckpts, we can not update this in this version, but for next, we will make a big change.

**Comment by Chen-Wendi** [comment]
Hi @wowfisch, @JinhuiYE.

I just created a PR #252 that adds an optional `LayerNorm` before action_decoder in QwenPI, which works well on my own dataset. I disabled the `LayerNorm` by default so that it will not affect any released ckpts.

Hope it helps.
